### PR TITLE
RFC 7233 (Range Request) Support (except for If-Range)

### PR DESCRIPTION
**Pull request by zlm2012**
I found that h2o does not support for range request for now so I just made a try to implement it. Tested via multithreaded downloading with aira2.

P.S. As a beginner of open source contributor, I know my implementation and coding behaviors may not be so good as other seniors. So any comment or guidance is welcomed. I'll try my best to improve it :-P

### CODE REVIEW

**Comment by kazuho** [comment]
Thank you for the PR!  Please wait a while until I look into the code.

**Comment by kazuho** [comment]
Sorry it took long for me to look into the PR.

The design looks generally fine to me.  I am happy to see that the changes are beautifully isolated from the normal code-path handling non-ranged requests, meaning that the degradation of performance would be minimum for ordinary workloads.

Let me leave some high-level comments.

* please add tests to t/00unit/lib/handler/file.t
 * tests should exist at both levels: a subtest that tests that the `process_range` function behaves as expected, and tests that test the range requests
 * FYI the entire test-suite can be run via `make check` (you may need to install additional perl modules, please consult `.travis.yml` to find out what is needed), or you can just build t-00unit-evloop.t and run it at the source directory
* it might be a good idea to first concentrate on polishing up the pushing generator (by commenting out the pulling generator)
* please define the length of the boundary using macro (so that it could be referred from other parts of the code) instead of using magic numbers

I will add in-line comments to other issues.

File: lib/handler/file.c [hunk 1]
```diff
@@ -46,6 +49,12 @@ struct st_h2o_sendfile_generator_t {
     char is_gzip;
     char send_vary;
     char *buf;
+    int range_count;
+    size_t *range_infos; /* size_t shows in pair. first is start offset, then length */
+    h2o_iovec_t boundary; /* boundary used for multipart/byteranges */
+    h2o_iovec_t mimetype; /* original mimetype for multipart */
+    int current_range; /* range that processing now */
+    int finished_range; /* range that processed till now */
```
Comment (kazuho) [inline lib/handler/file.c]: * please create an anonymous struct like `struct { ... } ranged;` for containing the properties necessary for handling ranged requests
* please use `size_t` for storing indices (my understanding is that `range_count`, `current_range`, `finished_range` all fit into the criteria)

File: lib/handler/file.c [hunk 2]
```diff
@@ -110,6 +119,46 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
...
+    size_t rlen, *range_cur = self->range_infos + 2 * self->finished_range, used_buf = 0;
+    ssize_t rret;
+    h2o_iovec_t vec;
+
+    if (self->bytesleft == 28) {
+        vec.len = sprintf(self->buf, "\r\n--%s--\r\n", self->boundary.base);
+        vec.base = self->buf;
+        h2o_send(req, &vec, 1, 1);
+        do_close(&self->super, req);
+        return;
```
Comment (kazuho) [inline lib/handler/file.c]: For the purpose of adding the _end-of-multipart_ marker, I would suggest allocating a new buffer from memory pool to store the marker, and pass it to h2o_send() at once with the last content (i.e. the second argument to the function should contain two iovecs, first one containing the content, and the second one containing the _end-of-multipart_ marker, and the third argument should be set to _2_ instead of _1_).

By doing that way the code would become simpler, and you would no longer to adjust `bytesleft`.

File: lib/handler/range.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /* most range requests contain only one range */
+    do {
+        range_start = -1; range_count = 0;
+        if (likely(*buf >= '0') && likely(*buf <= '9')) {
+            range_start = 0;
+            while (likely(*buf >= '0') && likely(*buf <= '9')) {
+                range_start *= 10;
+                range_start += *buf++ - '0';
+                CHECK_EOF();
+            }
```
Comment (kazuho) [inline lib/handler/range.c]: Please extract the logic of parsing a number to a dedicated function, and place it next to `h2o_strtosize` in lib/common/string.c.

I would be happy if the name and behavior of the function was modeled similar to `h2o_strtosize` with the extension that the end position of the string can be obtained.

File: lib/handler/range.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+#include "h2o.h"
+
+#if __GNUC__ >= 3
+# define likely(x) __builtin_expect(!!(x), 1)
+# define unlikely(x) __builtin_expect(!!(x), 0)
+#else
+# define likely(x) (x)
+# define unlikely(x) (x)
+#endif
```
Comment (kazuho) [inline lib/handler/range.c]: Please move the definitions to h2o/memory.h with changing there names to `H2O_LIKELY` and `H2O_UNLIKELY`.

File: lib/handler/range.c [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	    if (unlikely(range_start >= file_size)) {
+                good_range=0;
+	    }
+            if (unlikely(buf == buf_end)) {
+                range_count = file_size - range_start;
+                goto GotOneRange;
+            } else if (unlikely(*buf == ',')) {
+                buf++;
+                CHECK_EOF();
+                range_count = file_size - range_start;
```
Comment (kazuho) [inline lib/handler/range.c]: `bytes=` header must accept whitespaces after `,` (see RFC 7230 section 7)

File: lib/handler/range.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    char *buf = range_value->base, *buf_end = buf + range_value->len;
+    int good_range=1;
+    H2O_VECTOR(size_t) ranges = {};
+    
+    EXPECT_CHAR('b');
+    EXPECT_CHAR('y');
+    EXPECT_CHAR('t');
+    EXPECT_CHAR('e');
+    EXPECT_CHAR('s');
+    EXPECT_CHAR('=');
```
Comment (kazuho) [inline lib/handler/range.c]: simply doing `if (range_value->len < 6 || memcmp(range_value->base, "bytes=", 6) != 0)` might be sufficient

**Comment by zlm2012** [comment]
Thanks for the comments. I'll do them ASAP. For the pushing generator, I didn't really understood how it works (even now, sorry) so just tried writing as `do_procceed()`.

File: lib/handler/file.c (hunk 2 above)
Comment (zlm2012) [inline lib/handler/file.c]: Just to make sure I did not misunderstand, is this part what you meant "polishing up the push generator"? TBO I don't fully understand how `do_procceed()` work when file is larger than `buf`, though I just followed that function and write `do_multirange_proceed()`. What is the process after non-finished sending, i.e. when leave the proceed function after calling h2o_send with `is_finished` 0? Sorry for not fully understanding the workflow, I tried my best...

Comment (kazuho) [inline lib/handler/file.c]: > TBO I don't fully understand how `do_procceed()` work when file is larger than `buf`, though I just followed that function and write `do_multirange_proceed()`. What is the process after non-finished sending, i.e. when leave the proceed function after calling h2o_send with `is_finished` 0? Sorry for not fully understanding the workflow, I tried my best...

There is nothing for you to apologize.  I highly appreciate your contribution!

When `is_finished` is set to zero, the protocol handlers call the `proceed` callback of the generator when data passed in via `h2o_send` has been sent to the client.  This phase is repeated until the generator calls `h2o_send` with `is_finished` set to 1.

In other words, you should change the value of `generator->super.proceed` from `do_proceed` to `do_multipart_proceed` in case of a multi-part response (sorry I did not notice the issue when I read your code).

Regarding the issue bet. push and pull generators, the former is the default one which should be supported by all generators.  Pull generator is an _optimization_ which can be used only when the underlying procotol handlers / output streams provide support for the feature.  For example, [the chunked encoder only provides the pull interface](https://github.com/h2o/h2o/blob/00c44dbb3eb49a73d6222a8df77090804b736a3b/lib/handler/chunked.c#L89) (notice that only the `do_send` callback is registered, `start_pull` remains NULL).

So for the PR to work correctly you must implement the pushing generator appropriately, whereas the pulling generator being something that can be worked on as an improvement.  That is why I suggested concentrating on the pushing generator for the time being.

Comment (zlm2012) [inline lib/handler/file.c]: Understood. I will handle it :-)

**Comment by zlm2012** [comment]
Polished some functions and did what you commented. I'll add unit test soon.

**Comment by zlm2012** [comment]
Unit tests added. Though final unit test on multiple range failed in my environment since the behavior in testing is not the same as running. I tried using the same `Range` header requesting the same file (`1000.txt`). When running h2o standalone, the response is right. But in unit testing, it seems write the last part twice instead of writing the last part and end-of-multipart marker to the response (I used `printf()` to check it). This happens in `t-00unit-evloop.t`.

**Comment by kazuho** [comment]
Thank you for the updates.  I like the tests you added; they seem to cover a lot of corner cases.

I have fixed the issue in the test loopback you mentioned in commit 8fcfb67 (thank you for noticing!).  Please update to the newest master.  The test still fails, but it seems like a false positive reported by `check_multirange_body`.

By reading your tests, I have following suggestions.  I would appreciate it you could fix them.

* process_range function should:
 * accept (and ignore) preceding commas
 * spaces before comma
 * spaces should match `' '` (white space) _and_ `'\t'` (horizontal tab)
 * ref. RFC 7230 Section 3.2.3, Section 7, RFC 7233 Appendix D
 * for example, please add tests for lines like: `\t,1-3 ,, ,5-9,`
* at the handler level:
 * please move the handler-level tests added to `test_lib__handler__file_c` to a dedicated subtest
 * please check that `conn->body->bytes` == value of the `content-length` header
 * "\r\n" exists before the first boundary; while the spec. permits such empty lines (ref. RFC 7233 Appendix A, notes 1), I would appreciate it if it gets removed

**Comment by zlm2012** [comment]
Okay. I will handle them. For the empty line before the first boundary, I just followed how nginx do since at that time I'm not sure if my understanding of RFC is right.

And if these problem solved, may I add some configuration field so that user can choose whether range request should be supported or not? Also, should I consider streaming situation, i. e. when filesize part of `Content-Range` is `*`?

**Comment by kazuho** [comment]
> And if these problem solved, may I add some configuration field so that user can choose whether range request should be supported or not?

Would there be any reason to turn the feature off?  To me it seems like keeping the feature always on would be fine.

> Also, should I consider streaming situation, i. e. when filesize part of Content-Range is `*`?

No.  The file handler always knows the length of the content, and there is no reason to use `*`.

**Comment by zlm2012** [comment]
All should goes well now. More suggestions welcomed.

**Comment by kazuho** [comment]
Thank you for the fixes.  I highly appreciate them.  Leaving comments in-line.

File: lib/handler/range.c [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        return NULL;                            \
+    }
+
+#define CHECK_OVERFLOW(range)                   \
+    if (range == SIZE_MAX) {                    \
+        *ret = -1;                              \
+        return NULL;                            \
+    }
+
+size_t *process_range(h2o_mem_pool_t *pool, h2o_iovec_t *range_value, size_t file_size, size_t* ret)
```
Comment (kazuho) [inline lib/handler/range.c]: The function seems to fail in parsing the following patterns:
* `bytes=1-3 5-10` (IMO should be considered invalid per spec)
* `bytes=1-\t,5-10` (IMO should return two ranges)

To me it seems like that it is possible to fix the issues by making the logic simpler by doing the following:
* logic that skips `[ \t]*,[ \t]*` should exist only at the top of the do-while block
 * current code does some checks at the top, and others within the logic that extracts a single range
* rest of the do-while block should concentrate on extracting a single range, until it reaches EOF or an unknown character

File: lib/handler/file.c [hunk 8]
```diff
@@ -225,15 +358,25 @@ static void do_send_file(struct st_h2o_sendfile_generator_t *self, h2o_req_t *re
...
-
+    assert(fcntl(self->fd, F_GETFL)>=0);
+    if (self->ranged.range_count == 1) 
+        lseek(self->fd, self->ranged.range_infos[0], SEEK_SET);
     if (req->_ostr_top->start_pull != NULL) {
-        req->_ostr_top->start_pull(req->_ostr_top, do_pull);
+        if (self->ranged.range_count < 2)
+            req->_ostr_top->start_pull(req->_ostr_top, do_pull);
+        else
+            req->_ostr_top->start_pull(req->_ostr_top, do_multirange_pull);
```
Comment (kazuho) [inline lib/handler/file.c]: `do_multirange_full` function seems to generate incorrect output.  If I build and run the server and send a request with multiple ranges, the response includes two end-of-multipart lines (see https://gist.github.com/kazuho/e69eaebfb87bd3fa2577).

The issue is not caught by the unit-tests, since the loopback handler of the test library only provides support for push-based generators.

As I suggested in one of my previous comments, implementing the pulling generator is optional, and I believe for the time being it is better to remove the pulling generator for the time being.  If you agree, you can remove the code by applying [this git diff](https://gist.github.com/kazuho/e636eac01496eb3ea517).

File: lib/handler/file.c [hunk 9]
```diff
@@ -372,11 +533,68 @@ static int on_req(h2o_handler_t *_self, h2o_req_t *req)
...
+            generator->bytesleft = range_infos[1];
+        else {
+            generator->ranged.mimetype = h2o_strdup(&req->pool, mime_type.base, mime_type.len);
+            size_t final_content_len = 0, size_tmp = 0, size_fixed_each_part, i;
+            h2o_iovec_t end_marker;
+            end_marker.base = h2o_mem_alloc_pool(&req->pool, BOUNDARY_SIZE + 8 + 1);
+            generator->ranged.boundary.base = h2o_mem_alloc_pool(&req->pool, BOUNDARY_SIZE + 1);
+            generator->ranged.boundary.len = BOUNDARY_SIZE;
+            h2o_gen_rand_string(&generator->ranged.boundary);
+            end_marker.len = sprintf(end_marker.base, "\r\n--%s--\r\n", generator->ranged.boundary.base);
```
Comment (kazuho) [inline lib/handler/file.c]: please generate `end_marker` when it is used, within `do_multirange_proceed`.

File: lib/handler/file.c [hunk 10]
```diff
@@ -26,11 +26,18 @@
...
 #include <unistd.h>
+
 #include "h2o.h"
 
 #define MAX_BUF_SIZE 65000
+#define BOUNDARY_SIZE 20
+#define FIXED_PART_SIZE 49
+/* = size of five "\r\n"s + size of "--" 
+ *   + size of header names w/ column and space (Content-Range and Content-Type) 
+ *   + size of "bytes " and "-" and "/" */
```
Comment (kazuho) [inline lib/handler/file.c]: Could the definition be more expressive (instead of using comments)?

It would be great if you could write like `sizeof("\r\n--") - 1 + BOUNDARY_SIZE + sizeof("\r\nContent-Range: bytes=-/\r\nContent-Type: \r\n\r\n") - 1` so that it would be easier to understand.

File: lib/handler/file.c [hunk 11]
```diff
@@ -110,6 +127,57 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
...
 }
 
+static void do_multirange_proceed(h2o_generator_t *_self, h2o_req_t *req) {
+    struct st_h2o_sendfile_generator_t *self = (void *)_self;
+    size_t rlen, *range_cur = self->ranged.range_infos + 2 * self->ranged.finished_range, used_buf = 0;
+    ssize_t rret, vecarrsize;
+    h2o_iovec_t vec, *vecarr;
+    int is_finished;
+
+    if (self->ranged.current_range == self->ranged.finished_range) {
```
Comment (kazuho) [inline lib/handler/file.c]: This is kind of a nitpick, but I suggest to make the code simpler by doing the following:
* instead of adjusting the value of `*range_cur`, use `self->bytesleft` to store the number of remaining bytes to be sent within the range, and use `self->bytesleft == 0` to detect if it is necessary to emit the boundary
 * this has a nice side-effect that the variable is used the same way as with non-multirange responses

File: lib/handler/file.c [hunk 12]
```diff
@@ -110,6 +127,57 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
...
+        vecarr = (void *)h2o_mem_alloc_pool(&req->pool, 2 * sizeof(h2o_iovec_t));
+        *vecarr = vec;
+        *(vecarr + 1) = self->ranged.end_marker_buf;
+        vecarrsize = 2;
+        is_finished = 1;
+    } else {
+        vecarr = &vec;
+        vecarrsize = 1;
+        is_finished = 0;
+    }
```
Comment (kazuho) [inline lib/handler/file.c]: How about declaring a `vec` as `h2o_iovec_t [2]`, and always using it (with `vecarrsize` set to `1` or `2`) instead of using `vec`, `vecarr` and calling `h2o_mem_alloc_pool`?

File: lib/handler/file.c [hunk 13]
```diff
@@ -110,6 +127,57 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
...
+        if (H2O_LIKELY(self->ranged.current_range != 0))
+            used_buf = sprintf(self->buf, "\r\n--%s\r\nContent-Type: %s\r\nContent-Range: bytes %zd-%zd/%zd\r\n\r\n",
+                               self->ranged.boundary.base, self->ranged.mimetype.base, *range_cur, range_end,
+                               self->ranged.filesize);
+        else
+            used_buf = sprintf(self->buf, "--%s\r\nContent-Type: %s\r\nContent-Range: bytes %zd-%zd/%zd\r\n\r\n",
+                               self->ranged.boundary.base, self->ranged.mimetype.base, *range_cur, range_end,
+                               self->ranged.filesize);            
+        self->ranged.current_range ++;
+        lseek(self->fd, *range_cur, SEEK_SET);
```
Comment (kazuho) [inline lib/handler/file.c]: Please close the response if `lseek` fails, in the same way as when `read` fails.

File: lib/handler/range.c (hunk 7 above)
Comment (zlm2012) [inline lib/handler/range.c]: RFC 7233 defines that 
```
byte-range-set = *( "," OWS ) ( byte-range-spec /
suffix-byte-range-spec ) *( OWS "," [ OWS ( byte-range-spec /
suffix-byte-range-spec ) ] )
```
So I think I should
* check `,[ \t]*` at the top of the do-while block (and skip). If `[ \t]*` is skipped when last single range parsing ended and no `,` found, error occurred.
* skip `[ \t]*` at each exit from single range parsing

Comment (kazuho) [inline lib/handler/range.c]: Thank you for checking the spec.  I agree with your comments.

**Comment by zlm2012** [comment]
Bugs are fixed. Committed changes.

**Comment by kazuho** [comment]
Thank you for the hard work!  I think we are mostly done.  Leaving comments in-line.

File: lib/handler/file.c [hunk 14]
```diff
@@ -282,6 +381,23 @@ static int send_dir_listing(h2o_req_t *req, const char *path, size_t path_len, i
     return 0;
 }
 
+size_t *process_range(h2o_mem_pool_t *pool, h2o_iovec_t *range_value, size_t file_size, size_t* ret);
```
Comment (kazuho) [inline lib/handler/file.c]: Please change the meaning of the last variable to _the number of ranges found, if the function succeeds in parsing the input_, from _the number of ranges found, or error code if it fails_.

I believe that the design of the function has been borrowed from the functions of picohttpparser.  But in the case of this function we do not need to distinguish parse errors and empty input (or incomplete input).

The caller of the function can (and does) simply ignore the result if the function returns `NULL`, so changing the meaning of the last argument as suggested will not have any side effects.  OTOH it would simplify the design of the function.

File: lib/handler/range.c [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        *ret = -1;
+        return NULL;
+    }
+
+    buf += 6;
+    
+    /* most range requests contain only one range */
+    do {
+        while (1) {
+            if (*buf != ',') {
```
Comment (kazuho) [inline lib/handler/range.c]: Please assert `buf != buf_end` somewhere before here, or we would might buffer overrun (for example when the `range_value` is `{"bytes=", 6}`).

File: lib/handler/range.c [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    if (range == SIZE_MAX) {                    \
+        *ret = -1;                              \
+        return NULL;                            \
+    }
+
+size_t *process_range(h2o_mem_pool_t *pool, h2o_iovec_t *range_value, size_t file_size, size_t* ret)
+{
+    size_t range_start = -1, range_count = 0;
+    char *buf = range_value->base, *buf_end = buf + range_value->len;
+    int good_range = 1, ows_skipped = 0;
```
Comment (kazuho) [inline lib/handler/range.c]: Please change the name of `ows_skipped` to something like `needs_comma` to reflect the meaning of the flag.

File: lib/handler/range.c [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+            range_start = file_size - range_count;
+            if (H2O_UNLIKELY(*buf == '-')) {
+                *ret = -1;
+                return NULL;
+            }
+        } else {
+            *ret = -1;
+            return NULL;
+        }
+    GotOneRange:
```
Comment (kazuho) [inline lib/handler/range.c]: Please use else-ifs and flags instead of gotos in this case to simplify the code path.

File: lib/handler/range.c [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+            }
+            ows_skipped = 0;
+            buf ++;
+            while (H2O_UNLIKELY(*buf == ' ') || H2O_UNLIKELY(*buf == '\t')) {
+                buf ++;
+                CHECK_EOF();
+            }
+        }
+        if (H2O_UNLIKELY(buf == buf_end)) break;
+        range_start = -1; range_count = 0;
```
Comment (kazuho) [inline lib/handler/range.c]: Please use `SIZE_MAX` instead of `-1` when assigning ~0 to `size_t`.

I would also suggest removing variable `good_range`, ad using `range_start != SIZE_MAX` for determining if the parser has succeeded in extracting a valid range.

File: lib/handler/range.c [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+            if (H2O_UNLIKELY(range_count > file_size - 1))
+                range_count = file_size - 1;
+            if (H2O_UNLIKELY(range_start > range_count)) {
+                good_range = 0;
+            }
+            range_count -= range_start - 1;
+            if (H2O_UNLIKELY(*buf == '-')) {
+                *ret = -1;
+                return NULL;
+            }
```
Comment (kazuho) [inline lib/handler/range.c]: I believe you do not need to check if ``-`` exists at the end of a single range (which is BTW accessing the buffer without boundary checking) if you unconditionally set `ows_skipped` (suggested to be renamed to `needs_comma`) unconditionally at the end of the do-while loop.

**Comment by kazuho** [comment]
Apart from the comment on `lib/handler/file.c` (and the change of the signature of `process_range` function), the only leftover is the issues in the `process_range` function, which I have left in-line comments above.

After those issues get resolved, I would suggest doing the following, and after then I expect this PR will be ready to get merged to master.

* move the macros in range.c into the function body (since they are only used by a single function), and `#undef` them at the end of the function body
* delete the function-like macros that are called only once
* move `process_range` function into `lib/handler/file.c` and unlink `lib/handler/range.c`

**Comment by zlm2012** [comment]
Maybe all should be done now? And maybe Issue #153 can be solved (at least partially).

**Comment by kazuho** [comment]
Made some tweaks, added end-to-end tests and merged to master.

Thank you very much for the contribution!  I will add your names to the copyright before we release 1.3.

**Comment by dch** [comment]
:dancer: thankyou @kazuho & @zlm2012 for this great addition!
